ansible_base/rbac/models.py

```diff
@@ -0,0 +1,397 @@
+
+        to_add = []
+        for codename, ct_id, obj_id in expected_evaluations - existing_set:
+            to_add.append(RoleEvaluation(codename=codename, content_type_id=ct_id, object_id=obj_id, role=self))
+
+        return (to_delete, to_add)
+
+
+# COMPUTED DATA
+class RoleEvaluation(models.Model):
+    """
+    Cached data that shows what permissions an ObjectRole gives its owners
+    example:
+        ObjectRole 423 gives users execute access to job template 37
+
+    RoleAncestorEntry model in old AWX RBAC system is a direct analog
+
+    This is used to make permission evaluations via querysets returning object ids
+    the data in this table is created from the ObjectRole and RoleDefinition data
+      you should not interact with this table yourself
+      the only method that should ever write to this table is
+        compute_object_role_permissions()
+
+    In the above example, "ObjectRole 423" may be a role that grants membership
+    to a team, and that team was given permission to another ObjectRole.
+    """
+
+    class Meta:
+        app_label = 'dab_rbac'
+        verbose_name_plural = _('role_object_permissions')
+        indexes = [
+            models.Index(fields=["role", "content_type_id", "object_id"]),  # used by get_roles_on_resource
+            models.Index(fields=["role", "content_type_id", "codename"]),  # used by accessible_objects
+        ]
+        constraints = [models.UniqueConstraint(name='one_entry_per_object_permission_and_role', fields=['object_id', 'content_type_id', 'codename', 'role'])]
+
+    def __str__(self):
+        return (
+            f'RoleEvaluation(pk={self.id}, codename={self.codename}, object_id={self.object_id}, '
+            f'content_type_id={self.content_type_id}, role_id={self.role_id})'
+        )
+
+    def save(self, *args, **kwargs):
+        if self.id:
+            raise RuntimeError('RoleEvaluation model is immutable and only used internally')
+        return super().save(*args, **kwargs)
+
+    role = models.ForeignKey(
+        ObjectRole, null=False, on_delete=models.CASCADE, related_name='permission_partials', help_text=_("The object role that grants this form of permission")
+    )
+    codename = models.TextField(null=False, help_text=_("The name of the permission, giving the action and the model, from the Django Permission model"))
+    # NOTE: we awkwardly do not form these into a content_object, following from AWX practice
+    # this can be relaxed as we have comparative performance testing to confirm doing so does not affect permissions
+    content_type_id = models.PositiveIntegerField(null=False)
+    object_id = models.PositiveIntegerField(null=False)
+
+    def obj_perm_id(self):
+        "Used for in-memory hashing of the type of object permission this represents"
+        return (self.codename, self.content_type_id, self.object_id)
+
+    @staticmethod
+    def accessible_ids(cls, actor, codename, content_types=None):
```

Comment:
Why is this `actor` instead of `user`? Can you call this for teams as well?

Reply:
Come to think of it, how do users inherit permissions from teams?

Reply:
Yes, you can call it for teams, because it relies on the same name relationship for users/teams when it does `actor.has_roles.all()`. I prefer to avoid using this method in a team-centric way, but AWX has a good handful of corner cases that needed it. I reluctantly added support.

Users are a member of a team by virtue of being give a role that grants membership to that team. That role has to contains the "member_team" permission for this to happen. For every team that has a resource object-role, `RoleEvaluation` entries are duplicated with the `role` field being the team membership role.

To rephrase, the point of `RoleEvaluation` is to duplicate the codename/content_object data for every role that inherits those permissions.